activerecord/lib/active_record/relation/from_clause.rb

```diff
@@ -25,7 +25,7 @@ def empty?
       end
 
       def self.empty
-        new(nil, nil)
+        @empty ||= new(nil, nil).freeze
```

Comment:
The pain caused by this requirement is pretty minimal; this requirement quickly surfaces in tests, while a comment indicating that these objects are immutable by convention is impossible to enforce.